### ISSUE TITLE
ask for suggestions for better infercence results from personal dataset

### ISSUE THREAD

**Issue by zclecle2** [MIC-DKFZ/nnUNet#1793]
Dear Fabian Isensee and other members in Applied Computer Vision Lab (ACVL),

Thank you for developing the nnUnet2! It's easy to use following the guidance in the GitHub. For my own data, I tried several times of training with different input data and I couldn't get satisfied inference result. I am not sure what can be improved on my part. I am wondering if you could help giving some advice about this. And I am from life science background and I am not so familiar with deep learning. Looking forward to your reply! Much thanks!

I would like to separate the axons in tissue immuno-histochemistry staining sections from other objects and background, but mainly to isolate axons, other objects are not that important. After training and inference, I found that some results are good as labeled by red rectangles in the following image. While those labeled by blue rectangles should not be regarded as axons. You may download the original images attached for a closer look. I have tried to use a patch size of 120 pixels or 500 pixels. I have tried training with batch size with 160 images(patch size 120 pixels) with all semi-manual labels. And tried training with batch size of 14,000 images with 120 pixels patch size or of 700 images with 500 pixels patch size with some previous inferred results with corrected labels. And what I show is the latest training I used, which is better than what I previously had, but still, they contained some unsatisfactory results.

![Picture1](https://github.com/MIC-DKFZ/nnUNet/assets/95748462/ac4dbad1-3420-4ca5-a5f0-c3bb990bd3fb)
original data:
![test_150_0000-1](https://github.com/MIC-DKFZ/nnUNet/assets/95748462/f7154834-2e75-436a-8b58-9c8eb753fa69)
![test_150_0000-2](https://github.com/MIC-DKFZ/nnUNet/assets/95748462/46e29a35-72ac-43f4-a1ad-b5578ffe01d4)
![test_150-1](https://github.com/MIC-DKFZ/nnUNet/assets/95748462/6aaacf02-cbfa-4b73-8665-109dfc18a5ae)
![test_150-2](https://github.com/MIC-DKFZ/nnUNet/assets/95748462/7f9811e8-dd9c-4f59-98f7-54a453e19a8c)

[dataset.json](https://github.com/MIC-DKFZ/nnUNet/files/13318313/dataset.json)
The exact preprocessing, training, and inference commands I have used: 
nnUNetv2_plan_and_preprocess -d 086 --verify_dataset_integrity
nnUNetv2_train 086 2d 0 -device cuda --npz
nnUNetv2_train 086 2d 1 -device cuda --npz
nnUNetv2_train 086 2d 2 -device cuda --npz
nnUNetv2_train 086 2d 3 -device cuda --npz 
nnUNetv2_train 086 2d 4 -device cuda --npz
nnUNetv2_train 086 2d all -device cuda --npz
nnUNetv2_find_best_configuration 086
nnUNetv2_predict -d Dataset086_slc -i K:\KL\zcle\nnUNet\nnUNet_raw\Dataset086_slc\imagesTs -o K:\KL\zcle\nnUNet\nnUNet_results\Dataset086_slc\imagesTs_out -f  0 1 2 3 4 -tr nnUNetTrainer -c 2d -p nnUNetPlans
nnUNetv2_apply_postprocessing -i K:\KL\zcle\nnUNet\nnUNet_results\Dataset086_slc\imagesTs_out -o K:\KL\zcle\nnUNet\nnUNet_results\Dataset086_slc\imagesTs_out_PP -pp_pkl_file K:\KL\zcle\nnUNet\nnUNet_results\Dataset086_slc\nnUNetTrainer__nnUNetPlans__2d\crossval_results_folds_0_1_2_3_4\postprocessing.pkl -np 8 -plans_json K:\KL\zcle\nnUNet\nnUNet_results\Dataset086_slc\nnUNetTrainer__nnUNetPlans__2d\crossval_results_folds_0_1_2_3_4\plans.json

**Comment by Karol-G** [comment]
Dear @zclecle2,

Thank you for reaching out and for utilizing nnU-Net in your work.

Thank you already for the images provided.

However, to better assist you, please additionally provide unmodified tif/png images along with their corresponding predictions and ground truths that exemplify the issues you're encountering. These can be shared via cloud storage as azip archive; send the link to [karol.gotkowski@dkfz.de](mailto:karol.gotkowski@dkfz.de). Rest assured, all data will be handled with strict confidentiality and will be deleted after the necessary assessments.

Additionally, could you compile and send over the training logs and the progress.png for every fold? Please include these in a zip archive. This information is crucial for diagnosing the problem more accurately.

We look forward to receiving the requested data and are ready to support you in improving your nnU-Net experience.

Best regards,
Karol Gotkowski

**Comment by Karol-G** [comment]
Hey @zclecle2,

I looked at your data today and think that at least some of your problems originate from labeling inconsistencies and a too-weak signal-to-noise ratio. I attached a couple of images that show your GT segmentations. 

The label color coding is the following:
Axon: Brown
Vessels: Blue
Glia: Light Purple
Non-tissue: Dark Purple

Bounding boxe colors:
Red: Suspected labeling inconsistencies
Orange: Suspected too-weak signal-to-noise ratio


Regarding the labeling inconsistencies, I believe that sometimes the axon and vessel class are labeled interchangeably/inconsistently (red bounding boxes) and sometimes not at all. This in turn can confuse the model and increase the class uncertainty during inference. Being consistent for classes that can look similar is quite relevant here, I believe.

Regarding the too-weak signal-to-noise ratio, I believe in many cases the glia class cannot be predicted correctly as there is simply not enough signal (orange bounding boxes). The glia GT segmentation often looks randomly assigned.

Of course, I can be mistaken regarding the origin of your issues, but I would recommend to revistst the GT segmentations and ensure a consistent and high quality of the GT segmentations across images.

I hope I could help you.

Kind regards,
Karol

![Image](https://github.com/MIC-DKFZ/nnUNet/assets/3471895/1c79521b-2daf-4606-9e66-86cbd61fcd7a)
![Image](https://github.com/MIC-DKFZ/nnUNet/assets/3471895/c602ea6b-bc19-4999-941f-b559ac0b75b9)
![Image](https://github.com/MIC-DKFZ/nnUNet/assets/3471895/f8965605-3f15-4afc-b9a0-b4bf4d81f003)
![Image](https://github.com/MIC-DKFZ/nnUNet/assets/3471895/df87a356-dc0d-40b3-b2d3-af13e60f8505)
![Image](https://github.com/MIC-DKFZ/nnUNet/assets/3471895/2046251f-67ad-4365-b1bb-8fc26d9bee53)
![Image](https://github.com/MIC-DKFZ/nnUNet/assets/3471895/988f6ebc-3b61-4da3-92da-71543605435b)
![Image](https://github.com/MIC-DKFZ/nnUNet/assets/3471895/a768cb9a-7b47-4f45-9b41-067ff131d56b)

**Comment by zclecle2** [comment]
Thanks for your suggestions. And sorry for this late reply.  As you may
know, it is quite time-consuming to label the axons (value =1) manually.
When I first tried the nnUnet2, I used training data from manual labeling
and after training it had some major defects in terms of prediction. I
suspected that I didn't provide enough training samples because the
conditions for my sample could vary a lot. As you may see, in some
pictures, there are many blood vessels that greatly influence the model
prediction ability while in some conditions are not that bad.  So later I
added many more training data, which included some prediction results from
the previous training model which I manually corrected some major errors
for object assigning in terms of labels. In this case, some labeling may
not be ground truth labeling. After that, the prediction seemed to be much
better, while it still showed some errors in some pictures, for example,
some blood vessels were mistakenly predicted as axons. I guess that I still
need to add more new training data, especially from those which showed not
so that good predictions. Is it possible thet when I added some new
training data, I didn't need to train all the data(including old training
data and new training data) every time? Along with my increased training
data, the time needed for training also increased and currently, I need
around 5 days to train my data. Much appreciated your nice reply!

Karol Gotkowski ***@***.***> 于2023年11月21日周二 22:43写道：

**Comment by Karol-G** [comment]
Hey,

Sorry for the late reply but I was sick.
I recommend to always train on all the data and to ensure that the GT data is of high quality. Splitting the training into multiple stages with different amounts of training data within each stage (which is different to pretraining) will lead to the commen phenomen of _catastrophic forgetting_. This means that the model forgets the previous data it was trained on and will only perform good on the new data.

Kind regards,
Karol

**Comment by zclecle2** [comment]
Take care! Got it! Many thanks to you for your nice suggestions and
patience.

Karol Gotkowski ***@***.***> 于2023年12月11日周一 16:17写道：
